Add protocols to interpolation framework

knelli2 opened the pull request:
## Proposed changes

Adds and uses the following protocols in the interpolation framework

- InterpolationTargetTag
- ComputeTargetPoints
- ComputeVarsToInterpolate
- PostInterpolationCallback

This is to help standardize and better document the interpolation framework as part of #3772. Some of the docs for `InterpolationTarget` have been moved to these protocols. In the future I plan on adding a whole dev guide for the interpolation framework so that it's easy for new users to find and follow.

Also use the new `name()` function of an InterpolationTargetTag when printing out apparent horizon info. This way, if we have an InterpolationTargetTag that is templated on something, the name that is printed out is actually different (i.e. in `ControlSystem/ApparentHorzion/Measurements.hpp`)

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Use ComputeTargetPoints protocol

knelli2 committed:
Use ComputeVarsToInterpolate protocol

knelli2 committed:
Use PostInterpolationCallback protocol

knelli2 committed:
clang-format and fix typos

knelli2 committed:
Use InterpolationTargetTag protocol

knelli2 committed:
Remove unused struct in parallel interpolator test

knelli2 committed:
Add protocol for ComputeTargetPoints

knelli2 committed:
Add protocol for ComputeVarsToInterpolate

knelli2 committed:
Add protocol for PostInterpolationCallback

knelli2 committed:
Add protocol for InterpolationTargetTag

wthrowe commented on src/ParallelAlgorithms/Interpolation/Protocols/ComputeTargetPoints.hpp:
```diff
@@ -0,0 +1,90 @@
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+namespace Frame {
+struct Grid;
+struct Inertial;
+struct Distorted;
+}  // namespace Frame
```
Unused.

wthrowe commented on src/ParallelAlgorithms/Interpolation/Protocols/ComputeTargetPoints.hpp:
```diff
@@ -0,0 +1,90 @@
+        decltype(ConformingType::points(
+            std::declval<const db::DataBox<DummyTag>&>(),
+            std::declval<const tmpl::type_<DummyMetavariables>&>(),
+            std::declval<const ::Tags::Time&>()))>;
```
The function takes a *tag* as argument?

nilsvu replied:
No this should be `double` or `size_t` or whatever the temporal ID is.

How does this actually compile? The `ApparentHorizon` target retrieves things out of the DataBox, but there's only a `DummyTag` in there. I think the way to go in these situations where the compile-time interface is not well-defined is (1) document it and only test the things that can easily be tested, and (2) consider improving the compile-time interface in a follow-up (in this case, the expected `temporal_id` can be exposed as a type alias, but I don't have a great solution for the DataBox).

I'm also worried about increasing compile time if we instantiate functions with "dummy types" just to check protocol conformance. The instantiations probably don't work in the first place, but if they do, they are not used. So I think we should only instantiate functions or types in protocols with the types that are going to be used.

wthrowe replied:
I do not remember the exact details, but there are places where the definition of a template is not instantiated if the declaration is sufficient to determine some compile-time value, such as the return type of a function.  (As an extreme example, see `std::declval` itself, which has no definition.)  That's probably why this compiles with bad tags.  (This might be the concept "ODR used", but I'd have to look up the details to be sure.)

knelli2 replied:
@wthrowe oops my bad, this shouldn't take a tag, just a double. And yes @wthrowe's explanation is correct. These functions have templates that pick up on all the internal tags and such because those tags can be different for every ComputeTargetPoints like you mentioned @nilsvu. In all of these protocols that have a `static_assert` where I use `std::declval`, I'm checking for 3 things,

1. The conforming type has a member function with this name
2. This member function takes the correct number of arguments and that these arguments are the "correct" type, where "correct" doesn't mean all internals of every type are exactly what they would be when actually used. Just the outermost type is correct (including const, reference, gsl::not_null, etc...). Here, this means that the tags inside the DataBox get picked up by the template so they can be *anything* (like a DummyTag)
3. The return type of this member function is correct.

The fact that this is all compile time, and that we never reach into the internals of this member function in the protocol to check if it conforms, allow us to check its signature without knowing the proper types. So when @wthrowe says this compiles with "bad tags", I don't think that's accurate. It is a **feature** of protocols to be able to check function signatures that have templates and give them fake tags.

I could, in theory, expose enough type aliases everywhere in all of these structs to use the proper tags when checking the function signature. This would constrain the function signatures even more. This will be some extra work though, but is probably doable.

wthrowe replied:
By "bad" I just meant something that wouldn't work if you actually tried to use the thing.  I don't have a problem with how the test is being done.

wthrowe commented on src/ParallelAlgorithms/Interpolation/Protocols/ComputeTargetPoints.hpp:
```diff
@@ -0,0 +1,90 @@
+ * \details A struct conforming to the `ComputeTargetPoints` protocol must have
+ *
+ * - a type alias `is_sequential` that is either `std::true_type` or
+ *   `std::false_type` which indicates if an interpolations depends on previous
```
"an interpolations depends" -> "interpolations depend"

wthrowe commented on src/ParallelAlgorithms/Interpolation/Protocols/ComputeVarsToInterpolate.hpp:
```diff
@@ -0,0 +1,79 @@
+    template <typename T>
+    struct has_signature_1<
+        T, std::void_t<decltype(T::apply(
+               std::declval<const gsl::not_null<Variables<example_list>*>>(),
```
Include for `not_null`

wthrowe commented on src/ParallelAlgorithms/Interpolation/Protocols/ComputeVarsToInterpolate.hpp:
```diff
@@ -0,0 +1,79 @@
+ * \snippet Helpers/ParallelAlgorithms/Interpolation/Examples.hpp apply_v2
+ *
+ * This apply function fills the `vars_to_interpolate_to_target` type alias from
+ * the InterpolationTargetTag. Here `Dim` is `Metavariables::volume_dim`,
```
There is no `Dim` in the code, but there probably should be.

wthrowe commented on src/ParallelAlgorithms/Interpolation/Protocols/PostInterpolationCallback.hpp:
```diff
@@ -0,0 +1,126 @@
+    T, std::void_t<decltype(T::apply(
+           std::declval<const db::DataBox<DummyTag>&>(),
+           std::declval<const Parallel::GlobalCache<DummyMetavariables>&>(),
+           std::declval<const ::Tags::Time&>()))>> : std::true_type {
```
Again, a *tag*?

nilsvu replied:
same comment as for `ComputeTargetPoints`

knelli2 replied:
Yes this should just be a `double`, my bad.

wthrowe commented on tests/Unit/Helpers/ParallelAlgorithms/Interpolation/Examples.hpp:
```diff
@@ -105,4 +107,41 @@ struct ExampleComputeVarsToInterpolate
+      const gsl::not_null<Parallel::GlobalCache<Metavariables>*> /*cache*/,
+      const TemporalId& temporal_id) {
+    /// [callback_v1]
+    if (intrp::InterpolationTarget_detail::get_temporal_id_value(temporal_id) >
+        1.0) {
+      return true;
+    } else {
+      return false;
+    }
```
`return intrp::InterpolationTarget_detail::get_temporal_id_value(temporal_id) > 1.0;`

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/ComputeVarsToInterpolate.hpp:
```diff
@@ -0,0 +1,79 @@
+    template <typename T>
+    struct has_signature_1<
+        T, std::void_t<decltype(T::apply(
+               std::declval<const gsl::not_null<Variables<example_list>*>>(),
```
Here too the I don't quite understand how things can work with a `DummyTag` when `ah::ComputeHorizonVolumeQuantities` assumes a bunch of specific tags must or can be present. I think we can do little more than document what's expected at the moment, and think about how to improve things. In this spirit, please add some notes on `allowed_src_tags` and friends to the docs.

knelli2 replied:
See other comment about templates.

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/InterpolationTargetTag.hpp:
```diff
@@ -0,0 +1,123 @@
+namespace detail {
+// Both will have `::value = true` because we want the static assert in the
+// protocol to pass regardless of if the `compute_vars_to_interpolate` alias
+// exists if the conforming type. It's just that if the alias does exist,
```
(first) if -> in

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/InterpolationTargetTag.hpp:
```diff
@@ -0,0 +1,123 @@
+ * \brief A protocol for `InterpolationTargetTag`s that are used in the
+ * intrp::InterpolationTarget parallel component.
+ *
+ * \details A struct conforming to the `InterpolationTarget` protocol must
```
`InterpolationTargetTag`

I think we should eventually remove the "tag" in the name because it has little resemblance to what we call tags in the rest of the code, but that's out of scope for this PR.

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/InterpolationTargetTag.hpp:
```diff
@@ -0,0 +1,123 @@
+ * have
+ *
+ * - a type alias `temporal_id` that tells the interpolation target what values
+ *   of time to use.
```
Should this be the tag or its type? Please clarify. (I think it's the tag)

knelli2 replied:
The tag.

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/InterpolationTargetTag.hpp:
```diff
@@ -0,0 +1,123 @@
+ *   of time to use.
+ *
+ * - a static function `name()` that returns a `std::string` used to uniquely
+ *   identify this InterpolationTargetTag
```
You added this requirement in this PR, right? I suggest you make it optional and use `Options::name<>()` where needed (last commit), which falls back to the name of the struct. This avoids adding the `pretty_type` code that you added to every interpolation target. (we should probably move `Options::name` somewhere more general or merge it with `pretty_type`..).

knelli2 replied:
I have plans to do exactly that: replace `pretty_type::short_name` with `pretty_type::name` that does exactly what `Options::name` does. I'm getting tired of adding options to files that don't have anything to do with initialization just to use the `Options::name` function. For now I'd like to keep options out of the interpolation framework though (except where actually needed)

nilsvu replied:
`pretty_type::name` sounds good. For the moment though you only need to include the options header to use `Options::name`, no other option things necessary, so I think that's not too bad?

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/InterpolationTargetTag.hpp:
```diff
@@ -0,0 +1,123 @@
+ *
+ * - a type alias `compute_target_points` that conforms to the
+ *   intrp::protocols::ComputeTargetPoints protocol. See this protocol for
+ *   requirements on `compute_target_points`.
```
You can delete this last sentence. Perhaps add a note how the alias is used instead. (same comment for the entries below)

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/InterpolationTargetTag.hpp:
```diff
@@ -0,0 +1,123 @@
+ *   intrp::protocols::PostInterpolationCallback protocol. See this protocol
+ *   for requirements on `post_interpolation_callback`.
+ *
+ * A struct conforming to the `InterpolationTarget` protocol can also optionally
```
InterpolationTargetTag (or just "this")

nilsvu commented on src/ParallelAlgorithms/Interpolation/Targets/ApparentHorizon.hpp:
```diff
@@ -136,7 +138,7 @@ namespace TargetPoints {
 /// For requirements on InterpolationTargetTag, see InterpolationTarget
 template <typename InterpolationTargetTag, typename Frame>
-struct ApparentHorizon {
+struct ApparentHorizon : tt::ConformsTo<intrp::protocols::ComputeTargetPoints> {
```
Please add a `static_assert(tt::assert_conforms_to)` to the corresponding test for each of the structs you have added protocol conformance to. This will probably fail to compile for the reasons I commented above (DummyTags etc). If it compiles, I'd like to understand why.

knelli2 replied:
It will compile, see other comment.

nilsvu replied:
Ok good! The `assert_conforms_to` should be part of the tests nonetheless I think. Just a one-line static_assert somewhere.

nilsvu commented on src/ParallelAlgorithms/Interpolation/Callbacks/ObserveTimeSeriesOnSurface.hpp:
```diff
@@ -89,19 +91,19 @@ auto make_reduction_data(const db::DataBox<DbTags>& box, double time,
+  template <typename DbTags, typename Metavariables, typename TemporalId>
+  static void apply(const db::DataBox<DbTags>& box,
+                    Parallel::GlobalCache<Metavariables>& cache,
+                    const TemporalId& temporal_id) {
```
I'm unsure about this change, and suggest you drop it, along with the similar changes below (to `SendGhWorldtubeData` etc). This change shouldn't be needed to conform to the protocol. As I commented above, perhaps the right thing to do is to constrain the compile-time interface more, not less, but no need to decide in this PR.

knelli2 replied:
See other comment explaining how templates are used to deduce types in protocols.

nilsvu reviewed (changes requested):
This adds so much structure to the interpolation framework docs, amazin 👍

knelli2 commented:
Rebased and fixed conflicts. When fixing conflicts, I made the minimal changes necessary to the original commits to get everything to work.

@nilsvu To address your comments about constraining the compile-time interface better I had to standardize some of the structs used in the interpolation framework. These are what the "NEW COMMIT" commits are addressing. (Basically just making most things templated on the interpolation target tag, I'll edit the commit message when I squash). I thought about it a lot and don't believe it's possible to constrain the metavariables or the DataBox tags passed to these functions without a significant overhaul of the protocol infrastructure. If a function requires certain things in the metavars or tags, it can just `static_assert` them or use an `if constexpr` block.

Now that #3909 is merged, I removed the `name()` requirement for InterpolationTargetTag.

No longer need `observation_type` in PostInterpolationCallback courtesy of #3904.

I split all fixups into their corresponding commits so reviewers know what changes go with what commits.

nilsvu commented on src/ParallelAlgorithms/Interpolation/Targets/LineSegment.hpp:
```diff
@@ -119,26 +119,31 @@ namespace TargetPoints {
+/// \note This takes a `tmpl::size_t<Dim>` as its second template argument so
+/// that it is consistent with the other Targets having only type template
+/// parameters and no non-type template parameters.
+template <typename InterpolationTargetTag, typename VolumeDim>
```
I see why you did this, but I think we have to find another solution.. It seems impractical to limit template parameters to `typename` just so they're easier to handle in protocols. Brigand does this to an extent, and `::Tags::deriv` does this for compatibility with `db::wrap_tags_in`, but I've found it very annoying and meant to refactor that for a while now. I think it would be much too limiting to impose a rule like "if a class conforms to a protocol, it can only have `typename` template parameters". We should instead improve our tools to handle this situation.

knelli2 replied:
Yeah I wasn't a fan of doing this but it was necessary to get everything to compile.

It's not that protocols only require `typename` templates, it's just that when doing `template template` expressions, all template parameters have to be similar or you have to know you're always getting a `non-type` parameter so you can account for it. It's not specific to protocols.

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/ComputeTargetPoints.hpp:
```diff
@@ -63,24 +55,36 @@ namespace intrp::protocols {
+
+  template <template <typename...> class ConformingType,
+            typename InterpolationTargetTag, typename... Ts>
+  struct test<ConformingType<InterpolationTargetTag, Ts...>> {
```
There's actually an easy enough way around this I think: Either require the conforming type exposes the `InterpolationTargetTag` as a type alias, from which you can then retrieve what you need. Not sure if there's in issue with circular dependencies. Or, make the `temporal_id` a requirement of the `ComputeTargetPoints` protocol itself, so it doesn't have to "reach up" into the InterpolationTargetTag.

knelli2 replied:
I like avoiding the `template template` expressions, but this adds the additional requirement that `temporal_id` defined in the `InterpolationTargetTag` be exactly the same as the one defined in `ComputeTargetPoints` (otherwise bad things might happen). This is essentially equivalent to making `ComputeTargetPoints` templated on `InterpolationTargetTag` to ensure this...

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/PostInterpolationCallback.hpp:
```diff
@@ -116,11 +117,25 @@ struct has_signature_3<
+
+  template <template <typename...> class ConformingType,
+            typename InterpolationTargetTag, typename... Ts>
+  struct test<ConformingType<InterpolationTargetTag, Ts...>> {
```
Same comment about avoiding to "reach up" into the InterpolationTargetTag. I think you only need the temporal ID here.

nilsvu reviewed (commented):
Another comment on the protocol implementation below. For the record, I think it would be perfectly fine (and probably easier to review) if this PR only added the protocols with documentation, plus the subset of tests that are straightforward to do, without changing any implementation. If you want to go beyond that then I'm fine with it, but the number of decisions that need to be made can get quickly out of hand :) So doing these things in follow-up PRs could be easier.

knelli2 commented:
@nilsvu Yeah I agree that this PR has become quite complicated already. I'm perfectly fine changing this to only *add* the protocols, but this requires us to decide on an interface for each one.

Maybe @nilsvu @wthrowe we should have a meeting to discuss this rather than hashing it out in the comments?

nilsvu commented:
@knelli2 if you want to simplify the PR, you can add only the protocols with docs and the subset of conformance-tests that require no changes to the implementation.

knelli2 commented:
Discussed with @wthrowe and @nilsvu on 4/6/22, it was decided that protocols with functions that need a temporal ID must be able to take multiple types because a temproal ID can be several different types (`double`, `TimeStepID`, `LinkedMessageID`). Because of this, it didn't make much sense to constrain the compile time interface of the temporal ID. This significantly simplified the fixups necessary.

Current fixups address reviewers comments and still implement the following changes from before

> Now that https://github.com/sxs-collaboration/spectre/pull/3909 is merged, I removed the name() requirement for InterpolationTargetTag.
>
> No longer need observation_type in PostInterpolationCallback courtesy of https://github.com/sxs-collaboration/spectre/pull/3904.

nilsvu commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -211,6 +214,9 @@ struct EvolutionMetavars {
   };
 
+  static_assert(tt::assert_conforms_to<
+                SphericalSurface, intrp::protocols::InterpolationTargetTag>);
```
Please delete these asserts in the metavars. Protocol conformance is checked (1) where the interface is _used_, e.g. in the InterpolationTarget component, and (2) in unit tests.

nilsvu commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp:
```diff
@@ -319,18 +325,17 @@ struct EvolutionMetavars {
+              GeneralizedHarmonic::Tags::
+                  FourIndexConstraintCompute<3, ::Frame::Inertial>,
+              GeneralizedHarmonic::Tags::FConstraintCompute<3,
+                                                            ::Frame::Inertial>,
```
Generally try to reformat only the lines your are changing in a file, to help reviewing and to avoid unnecessary rebase conflicts. I'm not requesting you drop these changes, just keep it in mind please. You can always put a "clang-format this file" commit at the beginning if some formatting really annoys you ;)

knelli2 replied:
This is already in the `clang-format and fix typos` commit. The first one.

nilsvu replied:
Oh sorry, missed that 👍

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+  ComputeVarsToInterpolate.hpp
+  PostInterpolationCallback.hpp
+  ProtocolsNamespace.hpp
+  InterpolationTargetTag.hpp
```
alphabetize

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/ComputeVarsToInterpolate.hpp:
```diff
@@ -0,0 +1,116 @@
+               std::declval<const Jacobian<DataVector, 3, Frame::Grid,
+                                           Frame::Inertial>&>(),
+               std::declval<const InverseJacobian<
+                   DataVector, 3, Frame::ElementLogical, Frame::Grid>&>()))>>
```
Dim = 3 is hard-coded here. Is that intentional?

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/InterpolationTargetTag.hpp:
```diff
@@ -0,0 +1,119 @@
+ *   intrp::protocols::ComputeTargetPoints protocol. This will compute the
+ *   points on the surface that we are interpolating onto.
+ *
+ * - a type alias `post_interpolation_callback` that conforms to the
```
This is a list now, right?

knelli2 replied:
No that's `post_horizon_find_callbacks` from https://github.com/sxs-collaboration/spectre/pull/3902. Maybe eventually we can make this also a list?

nilsvu replied:
oh right!

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_SendGhWorldtubeData.cpp:
```diff
@@ -16,13 +16,20 @@
 
+// Doesn't have to be an actual interpolation target tag. Just needs a
+// temporal_id type alias
+struct FakeIntrpTargetTag {
```
You don't need this anymore I think

nilsvu commented on tests/Unit/ParallelAlgorithms/Interpolation/Test_InterpolationTargetLineSegment.cpp:
```diff
@@ -17,20 +17,27 @@
     using compute_target_points =
         ::intrp::TargetPoints::LineSegment<InterpolationTargetA, 3>;
+    using post_interpolation_callback =
+        intrp::callbacks::ObserveTimeSeriesOnSurface<tmpl::list<>,
```
When rebasing to make this a list, you can make the list empty for the tests that don't need a callback.

nilsvu reviewed (commented):
Looks good, please squash including these comments, except maybe keep the Dim=3 changes in a fixup if you decide to make changes for that.

knelli2 commented:
@nilsvu Edited the current fixups with your changes so @wthrowe has a chance to review them. I'll squash once he approves.

wthrowe commented on src/ParallelAlgorithms/Interpolation/Protocols/PostInterpolationCallback.hpp:
```diff
@@ -33,14 +31,14 @@ struct has_signature_1<
            std::declval<const db::DataBox<DummyTag>&>(),
            std::declval<const Parallel::GlobalCache<DummyMetavariables>&>(),
-           std::declval<const ::Tags::Time&>()))>> : std::true_type {
+           std::declval<const TimeStepId&>()))>> : std::true_type {
```
Don't use `TimeStepId` here.  It looks reasonable but is almost always wrong, so don't use it where someone might look for an example.

knelli2 replied:
Made it `double` here and in ComputeTargetPoints

wthrowe commented on src/ParallelAlgorithms/Interpolation/Protocols/InterpolationTargetTag.hpp:
```diff
@@ -44,44 +44,43 @@ constexpr bool if_alias_exists_assert_conforms_to_v =
- *   intrp::protocols::PostInterpolationCallback protocol. See this protocol
- *   for requirements on `post_interpolation_callback`.
+ *   intrp::protocols::PostInterpolationCallback protocol. After the
+ *   interpolation is complete, call this structs `apply` function.
```
"structs" -> "struct's"

wthrowe commented on tests/Unit/ControlSystem/Test_Measurement.cpp:
```diff
@@ -68,9 +70,16 @@ struct Metavariables {
-// the protocol examples.
+// control_system::RunCallbacks, does additional testing of
+// the protocol examples, and tests protocol conformity of
+// ApparentHorizons/Measurements.hpp.
```
Move this to an appropriate file.

knelli2 replied:
Added a new test for `BothHorizons`

nilsvu commented on src/ParallelAlgorithms/Interpolation/Callbacks/SendGhWorldtubeData.hpp:
```diff
@@ -25,12 +27,13 @@ namespace callbacks {
   static void apply(const db::DataBox<DbTags>& box,
                     Parallel::GlobalCache<Metavariables>& cache,
-                    const typename TemporalId::type& temporal_id) {
+                    const TemporalId& temporal_id) {
```
The `Cce::Actions::ReceiveGhWorldtubeData` simple action expects that the `temporal_id` is a `TimeStepId`. Any other type will cause the simple action invocation to give a runtime error. Therefore, I suggest you hardcode the type to `TimeStepId` here. The compile time error from invoking this function with the wrong type is probably easier to debug than the runtime error from the simple action invocation.

I realize that hardcoding the temporal ID type here conflicts with the protocol checking this function with a `double`. The only solution that I can think of at the moment is this:
1. Make an alias `temporal_id_type` part of the `PostInterpolationCallback` protocol and use it in the conformance check.
2. In the `InterpolationTargetTag` protocol, assert that `std::is_same_v<temporal_id::type, post_interpolation_callback::temporal_id_type>`.

@knelli2 and @wthrowe what do you think of this?

knelli2 replied:
If it expects a `TimeStepId` then it should just `static_assert(std:is_same_v<TemporalId, TimeStepId>)` in the `apply` function. I don't really like changing the nice general protocol to account for this one specific case. It shouldn't be the job of the protocol to account for specific use cases. Classes should conform to the protocol, and then if the class has any extra restrictions it should impose those itself.

Using this method of `static_assert` doesn't break the protocol because the protocol only looks at the function signature, and nothing about the internals. Does this seem like a good compromise?

nilsvu replied:
Ok, I'll have to think more if agree with your reasoning, but your suggestion seems like a reasonable solution for now 👍 Please add to the docs of `PostInterpolationCallback` that the apply function is expected to take any `TemporalId` type.

nilsvu commented on tests/Unit/ParallelAlgorithms/Interpolation/Test_InterpolationTargetVarsFromElement.cpp:
```diff
@@ -82,13 +83,12 @@ struct MockComputeTargetPoints
+  template <typename DbTags, typename Metavariables, typename TemporalId>
+  static void apply(const db::DataBox<DbTags>& box,
+                    const Parallel::GlobalCache<Metavariables>& /*cache*/,
+                    const TemporalId& temporal_id) {
```
Also hardcode the type to `TimeStepId` here, because the function below expects this type. No need to suggest the function is more generic than it is :)

nilsvu commented on tests/Unit/ParallelAlgorithms/Interpolation/Test_ParallelInterpolator.cpp:
```diff
@@ -143,11 +144,12 @@ struct TestFunctionHelper<Tags::Negate> {
   static void apply(const db::DataBox<DbTags>& box,
                     const Parallel::GlobalCache<Metavariables>& /*cache*/,
-                    const TimeStepId& /*temporal_id*/) {
+                    const TemporalId& /*temporal_id*/) {
```
Just commenting here to say this change is good: the function works with any temporal id, so the type can be generic.

nilsvu commented on src/ParallelAlgorithms/Interpolation/Protocols/ComputeTargetPoints.hpp:
```diff
@@ -0,0 +1,82 @@
+ *
+ * - a function `points` with the following signature
+ *
+ * \snippet Helpers/ParallelAlgorithms/Interpolation/Examples.hpp points
```
(optional) If you're changing other things, I suggest that you move this snippet to the bottom of these docs and expand it to show the entire example implementation of the protocol. Something like: "Here's an example of a class that conforms to this protocol: \snippet ...". I feel like that's more helpful and gives more context than small snippets of code. The same suggestion applies to the other places in this PR where you include snippets of individual function signatures.

nilsvu commented:
@markscheel this PR is close to getting merged, if you want to take a look.

markscheel commented on src/ParallelAlgorithms/Interpolation/Protocols/ComputeVarsToInterpolate.hpp:
```diff
@@ -0,0 +1,121 @@
+ *   SrcFrame is the frame of `Metavariables::interpolator_source_vars` and
+ *   TargetFrame is the frame of `vars_to_interpolate_to_target` from the
+ *   InterpolationTargetTag. The overload without Jacobians treats the case in
+ *   which TargetFrame is the same as SrcFrame. Used only *with* the
+ *   Interpolator ParallelComponent.
```
ComputeVarsToInterpolate will also be used without the Interpolator ParallelComponent after #3828

knelli2 replied:
Should I change the docs now? Or should I wait to see which is merged first?

nilsvu replied:
Common practice is that the PR that get merged second does the rebase. In this case it wouldn't hurt to just delete the sentence in this PR already, while you're updating the comment below. Then we can get this merged.

markscheel commented on src/ParallelAlgorithms/Interpolation/Protocols/InterpolationTargetTag.hpp:
```diff
@@ -0,0 +1,119 @@
+
+namespace intrp::protocols {
+namespace detail {
+// Both will have `::value = true` because we want the static assert in the
```
Unclear what 'Both' means.  Maybe "if_alias_exists_assert_conforms_to::value will always be true because..."

knelli2 commented:
Squashed @markscheel's changes directly (and rebased).

nilsvu commented:
Something went wrong, dox check failed.

Edit: I think the \snippet must be on a single line

knelli2 commented:
@nilsvu yes you're right. Unfortunately, when I auto-format the file it recognizes a line with 80+ characters and puts part of the snippet on a new line...

Edit: Ok now CI is happy with dox and formatting :)

nilsvu reviewed (approved):
Look good 👍 Thanks again for your work on this!

nilsvu commented:
@markscheel were your comments addressed to your satisfaction? :)
